### PR TITLE
[bucketing] custom_ops mode to hide inductor copies overhead

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #161499

Adding "_custom_ops" bucketing to temporary fallback to eager execution of for_each,
to workaround too many generated kernels on inductor side.

This PR also reverts parts of bucketing changes for cycles detection that resulted in accuracy problems.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

Differential Revision: [D81152293](https://our.internmc.facebook.com/intern/diff/D81152293)

### CODE REVIEW

**Comment by IvanKobzarev** [comment]
@IvanKobzarev has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D81152293).

File: torch/_inductor/fx_passes/bucketing.py [hunk 1]
```diff
@@ -261,6 +281,52 @@ def _rs_group_key(node: torch.fx.Node) -> tuple[str, str, torch.dtype]:
     )
 
 
+@torch.library.custom_op("bucketing::_pre_bucket_reduce_scatter", mutates_args={})
+def _pre_bucket_reduce_scatter(
+    rs_ins: list[torch.Tensor],
+    group_size: int,
+) -> torch.Tensor:
+    rs_ins_flattened = [x.view(group_size, -1) for x in rs_ins]
+    new_rs_in = torch.cat(rs_ins_flattened, dim=1).flatten()
```
Comment (wconstab) [inline torch/_inductor/fx_passes/bucketing.py:290]: How come the custom op hides overhead? Does this change how inductor lowers the cat?

**Review by eellison — APPROVED** [review]
